Add tutorial on integration testing

abaeyens opened the pull request:
This PR adds a tutorial on integration testing with `launch_testing` and complements the two existing tutorials on unit testing. This topic was first discussed in https://github.com/ros2/ros2_documentation/issues/4827 and thereafter on [ROS discourse](https://discourse.ros.org/t/towards-a-rosdoc-integation-testing-tutorial/40701/1). This is my first PR for the ROS 2 project, all feedback welcome!

Resolves https://github.com/ros2/ros2_documentation/issues/4827 and closes https://github.com/ros2/launch/issues/812.

abaeyens committed:
tutorial/integration_testing: first draft

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/towards-a-rosdoc-integation-testing-tutorial/40701/2

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+.. TestingIntegration:
+
```
Since this is a new tutorial that has nothing linking to it, we don't need this.

```suggestion
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+Prerequisites
+-------------
+Before starting this tutorial,
+it is recommended to have completed the following tutorials on launching nodes:
```
```suggestion
Before starting this tutorial, it is recommended to have completed the following tutorials on launching nodes:
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+Background
+----------
+Where unit tests focus on validating a very specific piece of functionality,
+integration tests go all the way in the other direction.
```
```suggestion
Where unit tests focus on validating a very specific piece of functionality, integration tests focus on validating the interaction between pieces of code.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+----------
+Where unit tests focus on validating a very specific piece of functionality,
+integration tests go all the way in the other direction.
+In ROS 2 this means launching a system of one or several nodes,
+for example the Gazebo simulator and the Nav2 navigation stack.
```
```suggestion
In ROS 2 this is often accomplished by launching a system of one or several nodes, for example the `Gazebo simulator <https://gazebosim.org/home>`__ and the `Nav2 navigation <https://github.com/ros-planning/navigation2.git>`__ stack.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+for example the Gazebo simulator and the Nav2 navigation stack.
+As a result, these tests are more complex both to set up and to run.
+
+A key aspect of ROS 2 integration testing is that nodes part of different tests
+shouldn't communicate with each other, even when run in parallel,
+which will be achieved here using a specific test runner
+that picks unique ROS domain id's.
```
```suggestion
A key aspect of ROS 2 integration testing is that nodes that are part of different tests
shouldn't communicate with each other, even when run in parallel.
This will be achieved here using a specific test runner that picks unique :doc:`ROS domain IDs <../../../Concepts/Intermediate/About-Domain-ID>`.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+Overview
+--------
+The main tool in use here is the
+`launch_testing <https://docs.ros.org/en/{DISTRO}/p/launch_testing/index.html>`_
+package
```
```suggestion
The main tool in use here is the `launch_testing <https://docs.ros.org/en/{DISTRO}/p/launch_testing/index.html>`_ package
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+`launch_testing <https://docs.ros.org/en/{DISTRO}/p/launch_testing/index.html>`_
+package
+(`launch_testing repository <https://github.com/ros2/launch/tree/{REPOS_FILE_BRANCH}/launch_testing>`_).
+This ROS-agnostic functionality allows to extend a classic Python launch file
+with both active tests (that run while the nodes are also running)
+and post-shutdown tests (which run once after all nodes have exited).
```
```suggestion
This ROS-agnostic functionality can extend a Python launch file
with both active tests (that run while the nodes are also running)
and post-shutdown tests (which run once after all nodes have exited).
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+``launch_testing`` relies on the Python standard module
+`unittest <https://docs.python.org/3/library/unittest.html>`_
+for the actual testing.
+Hence, preferably avoid mixing with ``pytest``.
+If you prefer using pytest for integration tests,
+you will want to use the package
+`launch_pytest <https://docs.ros.org/en/{DISTRO}/p/launch_pytest/index.html>`_
+(`launch_pytest repository <https://github.com/ros2/launch/tree/{REPOS_FILE_BRANCH}/launch_pytest>`_),
+though this tutorial will discuss using unittest.
```
This part isn't actually true; we have tests in the core which mix `launch_testing` with `pytest`, like https://github.com/ros2/ros2cli/blob/rolling/ros2lifecycle/test/test_cli.py .  So I'd just remove this whole bit:

```suggestion
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+(`launch_pytest repository <https://github.com/ros2/launch/tree/{REPOS_FILE_BRANCH}/launch_pytest>`_),
+though this tutorial will discuss using unittest.
+
+To get our integration tests run as part of ``colcon test``,
+we register the launch file in the ``CMakeLists.txt``.
+To finish, we will briefly touch
+on how to run these tests and inspect the test results.
```
```suggestion
To get our integration tests run as part of ``colcon test``, we register the launch file in the ``CMakeLists.txt``.
To finish, we will briefly touch on how to run these tests and inspect the test results.
```

Besides this, it might also be good to discuss how this works for pure-Python packages as well.

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+1 Describe the test in the test launch file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Both the nodes under test and the tests themselves are launched
+using a Python launch file, which resembles a classic ROS 2 Python launch file.
+It is custom to let the integration test launch file names
+follow the pattern ``test/test_*.py``.
```
```suggestion
Both the nodes under test and the tests themselves are launched using a Python launch file, which resembles a ROS 2 Python launch file.
It is customary to make the integration test launch file names follow the pattern ``test/test_*.py``.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+It is custom to let the integration test launch file names
+follow the pattern ``test/test_*.py``.
+
+There are two common types of tests in integration testing:
+active tests, which run while the nodes under test are running,
+and post-shutdown tests, which are run after exiting the nodes.
+We will cover both in this tutorial.
```
```suggestion
There are two common types of tests in integration testing: active tests, which run while the nodes under test are running, and post-shutdown tests, which are run after exiting the nodes.
We will cover both in this tutorial.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+1.1 Imports
+~~~~~~~~~~~
+The top imports the Python modules we will be using.
+Only two modules are specific to testing:
+the general-purpose ``unittest``, and ``launch_testing``.
```
```suggestion
We first start by importing the Python modules we will be using.
Only two modules are specific to testing: the general-purpose ``unittest``, and ``launch_testing``.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+1.2 Generate the test description
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The function ``generate_test_description`` describes what to launch,
+similar to ``generate_launch_description``
+in a classic ROS 2 Python launch file.
```
```suggestion
The function ``generate_test_description`` describes what to launch, similar to ``generate_launch_description``
in a ROS 2 Python launch file.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+The function ``generate_test_description`` describes what to launch,
+similar to ``generate_launch_description``
+in a classic ROS 2 Python launch file.
+In the example below, we launch the turtlesim node
+and half a second later our tests.
```
```suggestion
In the example below, we launch the turtlesim node and half a second later our tests.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+In the example below, we launch the turtlesim node
+and half a second later our tests.
+
+In more complex integration test setups, you will probably want
+to launch a system of several nodes, together with additional nodes
+that performing mocking or must otherwise interact with the nodes under test.
```
```suggestion
In more complex integration test setups, you will probably want to launch a system of several nodes, together with additional nodes
that performing mocking or must otherwise interact with the nodes under test.
```

kscottz replied:
```suggestion
In more complex integration test setups, you will probably want
to launch a system of several nodes, together with additional nodes
that perform mocking or must otherwise interact with the nodes under test.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+In more complex integration test setups, you will probably want
+to launch a system of several nodes, together with additional nodes
+that performing mocking or must otherwise interact with the nodes under test.
+Including existing (XML) launch files - to avoid launch file duplication -
+is supported.
```
Can you explain more about what this sentence means?  The rest of this tutorial talks about Python launch files, so for consistency I think we should probably stick to Python launch files here.

abaeyens replied:
Removed in 5f46b89d24ef, it's not that relevant in this tutorial and is likely to confuse the reader.

What I meant with this: Instead of listing all nodes here, which is likely mostly a repetition of an already existing launch file, we can also include an existing Python, XML or YAML launch file. For example using `XMLLaunchDescriptionSource`:
```python3
            [
                # System under test
                IncludeLaunchDescription(
                    XMLLaunchDescriptionSource([os.path.join(
                        get_package_share_directory('app'),
                        'launch/turtles.launch.xml')])),
                # Launch tests 0.5 s later
                launch.actions.TimerAction(
                    period=0.5, actions=[launch_testing.actions.ReadyToTest()]),
            ]
        ),
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+1.3 Active tests
+~~~~~~~~~~~~~~~~
+The active tests interact with the running nodes. In this tutorial,
+we will check whether the Turtlesim node publishes pose messages
+(by listening to the node's 'turtle1/pose' topic)
+and whether it logs that it spawned the turtle
+(by listening to stderr).
```
```suggestion
The active tests interact with the running nodes.
In this tutorial, we will check whether the Turtlesim node publishes pose messages (by listening to the node's 'turtle1/pose' topic)
and whether it logs that it spawned the turtle (by listening to stderr).
```

kscottz replied:
@abaeyens we are not consistent in the capitalization of `turtlesim`. To remain consistent with the other docs we should probably go with no capitalization.

kscottz replied:
```suggestion
The active tests interact with the running nodes. In this tutorial,
we will check whether the turtlesim node publishes pose messages
(by listening to the node's 'turtle1/pose' topic)
and whether it logs that it spawned the turtle
(by listening to stderr).
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+from `unittest.TestCase <https://docs.python.org/3/library/unittest.html#unittest.TestCase>`_.
+The child class, here ``TestTurtleSim``, contains the following methods:
+
+- ``test_*``:
+  the test methods, each performing some ROS communication
+  with the nodes under test and/or listening to the process output
+  (passed in through ``proc_output``).
+  They are executed sequentially.
```
```suggestion
- ``test_*``: the test methods, each performing some ROS communication with the nodes under test and/or listening to the process output (passed in through ``proc_output``).
  They are executed sequentially.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  with the nodes under test and/or listening to the process output
+  (passed in through ``proc_output``).
+  They are executed sequentially.
+- ``setUp``, ``tearDown``:
+  respectively run before (to prepare the test fixture)
+  and after executing each test method.
+  By creating the node in the ``setUp`` method,
+  we use a different node instance for each test
+  to reduce the risk of tests contaminating each other.
```
```suggestion
- ``setUp``, ``tearDown``: respectively run before (to prepare the test fixture) and after executing each test method.
  By creating the node in the ``setUp`` method, we use a different node instance for each test to reduce the risk of tests communicating with each other.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  By creating the node in the ``setUp`` method,
+  we use a different node instance for each test
+  to reduce the risk of tests contaminating each other.
+- ``setUpClass``, ``tearDownClass``:
+  these class methods respectively run once before and after
+  executing all the test methods.
```
```suggestion
- ``setUpClass``, ``tearDownClass``: these class methods respectively run once before and after executing all the test methods.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+              'Spawning turtle [turtle1] at x=',
+              timeout=5, stream='stderr')
+
+Note that the way we listen to the 'turtle1/pose' topic
+in ``test_publishes_pose`` differs from
+:doc:`the usual approach <../../Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber>`.
```
```suggestion
Note that the way we listen to the 'turtle1/pose' topic in ``test_publishes_pose`` differs from :doc:`the usual approach <../../Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber>`.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+Note that the way we listen to the 'turtle1/pose' topic
+in ``test_publishes_pose`` differs from
+:doc:`the usual approach <../../Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber>`.
+Instead of calling the blocking ``rclpy.spin``,
+we trigger the ``spin_once`` method -
+which executes the first open callback
+(so, our subscriber callback if a message arrived within 1 s) -
+until we have gathered all messages published over the last 10 s.
```
```suggestion
Instead of calling the blocking ``rclpy.spin``, we trigger the ``spin_once`` method - which executes the first available callback (our subscriber callback if a message arrived within 1 second) - until we have gathered all messages published over the last 10 seconds.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+(so, our subscriber callback if a message arrived within 1 s) -
+until we have gathered all messages published over the last 10 s.
+
+If you want to go further, you can implement yourself a third test
+that publishes a twist message, asking the turtle to move,
+and subsequently checks that it moved
+by asserting that the pose message changed,
+effectively automating part of the
+`Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`.
```
```suggestion
If you want to go further, you can implement a third test that publishes a twist message, asking the turtle to move, and subsequently checks that it moved by asserting that the pose message changed,
effectively automating part of the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+1.4 Post-shutdown tests
+~~~~~~~~~~~~~~~~~~~~~~~
+The classes marked with the ``launch_testing.post_shutdown_test`` decorator
+are run after letting the nodes under test exit.
+A typical test here is whether the nodes exited cleanly,
+for which ``launch_testing`` provides the method
```
```suggestion
The classes marked with the ``launch_testing.post_shutdown_test`` decorator are run after letting the nodes under test exit.
A typical test here is whether the nodes exited cleanly, for which ``launch_testing`` provides the method
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+2 Register the test in the CMakeLists.txt
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Registering the test in the ``CMakeLists.txt`` fulfills two functions:
+it integrates it in the ``CTest`` framework ROS 2 CMake-based packages rely on
+(and hence it will be called when running ``colcon test``),
+and it also allows to specify *how* the test is to be run -
+in this case, with a unique domain id to ensure test isolation.
```
This sentence is too long; suggest to split this into two sentences (or two bullet points).

abaeyens replied:
Put it in bullet points in 5f46b89d24ef.

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+in this case, with a unique domain id to ensure test isolation.
+This latter aspect is realized using the special test runner
+`run_test_isolated.py <https://github.com/ros2/ament_cmake_ros/blob/{REPOS_FILE_BRANCH}/ament_cmake_ros/cmake/run_test_isolated.py>`_.
+To ease adding several integration tests,
+we define the CMake function ``add_ros_isolated_launch_test``
+such that each additional test requires only a single line.
```
```suggestion
To ease adding several integration tests, we define the CMake function ``add_ros_isolated_launch_test`` such that each additional test requires only a single line.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+3 Dependencies and package organization
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Finally, to avoid surprises,
+add the following dependencies to your ``package.xml``:
```
```suggestion
Finally, add the following dependencies to your ``package.xml``:
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  <test_depend>turtlesim</test_depend>
+
+
+After following the above steps, your package (here named 'app')
+ought to look as follows:
```
```suggestion
After following the above steps, your package (here named 'app') ought to look as follows:
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+    tests/
+        test_integration.py
+
+Concerning package organization:
+Integration tests can be part of any ROS package.
+One can dedicate one or more packages to just integration testing,
+or alternatively add them to the package of which they test the functionality.
+In this tutorial, we go with the first option
+as we will test the existing Turtlesim node.
```
```suggestion
Integration tests can be part of any ROS package.
One can dedicate one or more packages to just integration testing, or alternatively add them to the package of which they test the functionality.
In this tutorial, we go with the first option as we will test the existing Turtlesim node.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+4.1 Running with colcon
+~~~~~~~~~~~~~~~~~~~~~~~
+Running all tests is straightforward: simply run
+:doc:`colcon test <../../Intermediate/Testing/CLI>`.
+This command suppresses the test output
+and exposes little about which tests succeed and which fail.
+Useful therefore while developing tests is the option
+to print all test output while the tests are running:
```
```suggestion
Running all tests is straightforward: simply run :doc:`colcon test <../../Intermediate/Testing/CLI>`.
This command suppresses the test output and exposes little about which tests succeed and which fail.
Therefore while developing tests the ``--event-handlers`` option is useful to print all test output while the tests are running:
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+  Summary: 4 tests, 0 errors, 2 failures, 0 skipped
+
+lists two files:
+one ctest-formatted XML file (a result of the ``CMakeLists.txt``) and,
+more interestingly, also an XUnit-formatted XML file.
+This latter is suitable for automatic report generation
+in automated testing in CI/CD pipelines.
+If we would have also added unit tests, their XUnit files
+would show up as well here.
```
```suggestion
This command lists two files:
* a ctest-formatted XML file (a result of the ``CMakeLists.txt``)
* an XUnit-formatted XML file, which is suitable for automatic report generation in automated testing in CI/CD pipelines.

If we would have also added unit tests, their XUnit files would show up as well here.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+If we would have also added unit tests, their XUnit files
+would show up as well here.
+
+A suitable tool to visualize them all together is the
+`NodeJS package Xunit Viewer <https://github.com/lukejpreston/xunit-viewer>`_.
+It converts the XUnit files to HTML or straight into the terminal.
+For example, command and response (without highlighting):
+
+.. code-block:: console
+
+  $ xunit-viewer -r build/app/test_results -c
+    app.test_integration.launch_tests
+        ✗ test_publishes_pose time=0.52
+          - Traceback (most recent call last):
+    File "/home/user/ros_workspace/src/app/test/test_integration.py", line 67, in test_publishes_pose
+        assert len(msgs_rx) > 100
+            ^^^^^^^^^^^^^^^^^^
+    AssertionError
+        ✓ test_exit_codes time=0.0
+        ✓ test_logs_spawning time=0.197
+
+  1 failure, 2 passed
+  Written to: /home/user/ros_workspace/index.html
```
I'm actually going to suggest that we don't have this bit in here, because we haven't suggested to install this earlier.

Ryanf55 replied:
I think this can go in its own tutorial because it applies to any kind of tests, including unit tests too!

kscottz replied:
I would go with @Ryanf55 here and say that we link it as a follow on tutorial at the end.

abaeyens replied:
Shall we add it to the existing running tests tutorial? Proposal in https://github.com/ros2/ros2_documentation/pull/4881/commits/05f16a7cc7d5f280dd6bc119a5d5410fe0179138.

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+Summary
+-------
+
+In this tutorial, we explored the process of creating and running
+integration tests on the ROS 2 Turtlesim node. 
+We discussed the integration test launch file
+and covered writing active tests and post-shutdown tests.
+To recap, the four key elements of the integration test launch file are:
```
```suggestion
In this tutorial, we explored the process of creating and running integration tests on the ROS 2 Turtlesim node. 
We discussed the integration test launch file and covered writing active tests and post-shutdown tests.
To recap, the four key elements of the integration test launch file are:
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+and covered writing active tests and post-shutdown tests.
+To recap, the four key elements of the integration test launch file are:
+
+* The function ``generate_test_description``:
+  same as the classic way of launching nodes
+  (basically, it replaces ``generate_launch_description``).
+  It launches our nodes under tests as well as our tests.
```
```suggestion
* The function ``generate_test_description``: This launches our nodes under tests as well as our tests.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  same as the classic way of launching nodes
+  (basically, it replaces ``generate_launch_description``).
+  It launches our nodes under tests as well as our tests.
+* ``launch_testing.actions.ReadyToTest()``:
+  alerts the test framework that the tests should be run.
+  This ensures that the active tests and the nodes are run synchronously.
```
```suggestion
* ``launch_testing.actions.ReadyToTest()``: This alerts the test framework that the tests should be run, and ensures that the active tests and the nodes are run together.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+* ``launch_testing.actions.ReadyToTest()``:
+  alerts the test framework that the tests should be run.
+  This ensures that the active tests and the nodes are run synchronously.
+* An undecorated class inheriting from ``unittest.TestCase``:
+  houses the active tests, including set up and teardown.
+  One has access to the ROS logging through ``proc_output``.
```
```suggestion
* An undecorated class inheriting from ``unittest.TestCase``: This houses the active tests, including set up and teardown, and gives access to ROS logging through ``proc_output``.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+* An undecorated class inheriting from ``unittest.TestCase``:
+  houses the active tests, including set up and teardown.
+  One has access to the ROS logging through ``proc_output``.
+* A second class inheriting from ``unittest.TestCase``,
+  decorated with ``@launch_testing.post_shutdown_test()``.
+  As the name implies, these tests run after all nodes have shutdown.
+  A common assert here is to check the exit codes,
+  to ensure all nodes exited cleanly.
```
```suggestion
* A second class inheriting from ``unittest.TestCase`` decorated with ``@launch_testing.post_shutdown_test()``: These are tests that run after all nodes have shutdown; it is common to assert that the nodes exited cleanly.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  A common assert here is to check the exit codes,
+  to ensure all nodes exited cleanly.
+
+The launch test is subsequently registered in the ``CMakeLists.txt``
+using the custom cmake macro ``add_ros_isolated_launch_test``
+that ensures that each launch test runs with a unique ``ROS_DOMAIN_ID``,
+avoiding undesired cross communication.
```
```suggestion
The launch test is subsequently registered in the ``CMakeLists.txt`` using the custom cmake macro ``add_ros_isolated_launch_test`` which ensures that each launch test runs with a unique ``ROS_DOMAIN_ID``,
avoiding undesired cross communication.
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+that ensures that each launch test runs with a unique ``ROS_DOMAIN_ID``,
+avoiding undesired cross communication.
+
+To finish, tools such as Xunit Viewer ease visualizing the test results
+in a more colorful way than the colcon utilities.
+We hope this tutorial gave the reader a good grasp
+of how to conduct integration tests in ROS 2,
+delineating tests, and analyzing their results.
```
```suggestion
```

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+Next steps
+----------
+* Extend the two basic active tests with one that checks
+  whether the turtle is responsive to twist commands,
+  and another that verifies whether the spawn service
+  sets the turtle to the intended pose
+  (see the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`).
+* Instead of Turtlesim, launch the
+  :doc:`Gazebo simulator <../../Advanced/Simulators/Gazebo/Gazebo>`
+  and simulate *your* robot in there, automating tests
+  that would otherwise depend on manually operating your physical robot.
+* Go through the
+  `launch_testing documentation <https://docs.ros.org/en/{DISTRO}/p/launch_testing/index.html>`_
+  and explore the many possibilities for integration testing it offers.
```
This is basically "homework", and while I don't necessarily disagree with it, this isn't what we do in our other tutorials.  I would suggest just removing this for now.

abaeyens replied:
Ah, I see. The 'Next steps' sections in other tutorials point explicitly to other tutorials, while this is indeed more "homework". Removed in https://github.com/ros2/ros2_documentation/pull/4881/commits/07e941549049e2be0a1f207272f844ab2cad759e.

clalancette reviewed (changes requested):
First of all, thank you for the tutorial.  This is documentation we really need.

I've left quite a number of comments inline.  Please take a look when you get a chance.

Ryanf55 commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+
+      def test_publishes_pose(self, proc_output):
+          """Check whether pose messages published"""
+          msgs_rx = []
```
This pattern is what WaitForTopics attempts to help simplify. 
https://github.com/ros2/launch_ros/issues/348

abaeyens replied:
Thanks for pointing this out, I'll look into that.

abaeyens replied:
Mentioned the `launch_testing_ros` package and `WaitForTopics` in c28db38b.

Ryanf55 reviewed (commented):
Nice to see this tutorial written up. How do you deal with the DDS matching of topics taking longer than expected and causing test failures?

Timple commented:
Great initiative!

Is there any reason you chose the `unittest` framework over the `pytest` framework?

In my experience the `pytest` framework is much more complete. And the integration with `launch_pytest` using fixtures works fine.

The README there even suggests it's superior to `launch_testing`:

<blockquote>
launch_testing is an standalone testing tool, which lacks many features:

- It's impossible to filter test cases by name and run only some.
- It's impossible to mark a test as skipped or xfail.
- The error reporting of the tool was custom, and the output wasn't as nice as the output generated by other testing frameworks such as unittest and pytest.
</blockquote>

So if we're going to write an intetgration testing tutorial, I'd start off by recommending the recommended tool :slightly_smiling_face:

clalancette commented:
> Is there any reason you chose the `unittest` framework over the `pytest` framework?

While I agree with you that it would be nice to also have documentation on the `launch_pytest` framework, at least the core still heavily uses the `unittest` framework.  Thus, I'm fine with taking this documentation and then expanding it later to include the `pytest` one.

kscottz commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+-------
+
+In this tutorial, we explored the process of creating and running
+integration tests on the ROS 2 Turtlesim node. 
```
```suggestion
integration tests on the ROS 2 turtlesim node. 
```

kscottz commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  and another that verifies whether the spawn service
+  sets the turtle to the intended pose
+  (see the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`).
+* Instead of Turtlesim, launch the
```
```suggestion
* Instead of turtlesim, launch the
```

kscottz commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  whether the turtle is responsive to twist commands,
+  and another that verifies whether the spawn service
+  sets the turtle to the intended pose
+  (see the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`).
```
```suggestion
  (see the `turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`).
```

kscottz commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+  sets the turtle to the intended pose
+  (see the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`).
+* Instead of Turtlesim, launch the
+  :doc:`Gazebo simulator <../../Advanced/Simulators/Gazebo/Gazebo>`
```
I am not sure this is necessary. The link is just an intro Gazebo and is a bit unsatisfying. If we had a tutorial on running integration tests with Gazebo that would be worth including.

abaeyens replied:
Removed this section as part of https://github.com/ros2/ros2_documentation/pull/4881#discussion_r1865056188.

kscottz reviewed (approved):
I don't have much to add here that @clalancette hasn't already said. There is some inconsistency with the capitalization of `turtlesim` that needs to be resolved. Assuming the suggestions are merged, and the latter section on color rendering gets broken out to another tutorial (this could be a subsequent pull request), this is good enough to be merged.

fujitatomoya commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,432 @@
+Writing Basic Integration Tests with launch_testing
+===================================================
+
+**Goal:** Create and run integration tests on the ROS 2 Turtlesim node.
```
```suggestion
**Goal:** Create and run integration tests on the ROS 2 turtlesim node.
```

fujitatomoya reviewed (commented):
besides the comments, linter/doc8 are not happy https://github.com/ros2/ros2_documentation/actions/runs/12104682983/job/33807471087?pr=4881. can you also address those errors?

abaeyens commented:
@clalancette
Greatly appreciated that you took your time to review thoroughly. One question related to style, I see that a few of your suggestions (for example [this one](https://github.com/ros2/ros2_documentation/pull/4881/files#r1865046599)) remove linebreaks and don't change the content, could you briefly elaborate on the reason behind this?

As-is, I wrapped the lines to 80 characters allowing for clean git diffs. [This style guide suggest wrapping, but doesn't mention a desired line length for documentation](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#markdown-restructured-text-docblocks). Is a 100-character limit preferred for documentation?

@Ryanf55
> How do you deal with the DDS matching of topics taking longer than expected and causing test failures?

Such things can happen, though I don't immediately see how this relates specifically to testing. The DDS taking too long to match topics can also arise when deployed, in which case it's a bug, and therefore it's desirable that the integration tests point out that part of our system is failing?

@kscottz
Concerning the section on visualizing test output (https://github.com/ros2/ros2_documentation/pull/4881#discussion_r1866471220, https://github.com/ros2/ros2_documentation/pull/4881/files#r1865054539), shall we make it part of the existing tutorial [Running Tests in ROS 2 from the command line](https://docs.ros.org/en/jazzy/Tutorials/Intermediate/Testing/CLI.html)? Or do you prefer a short but separate tutorial? Proposal in https://github.com/ros2/ros2_documentation/pull/4881/commits/05f16a7cc7d5f280dd6bc119a5d5410fe0179138.

clalancette commented:
> As-is, I wrapped the lines to 80 characters allowing for clean git diffs. [This style guide suggest wrapping, but doesn't mention a desired line length for documentation](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#markdown-restructured-text-docblocks). Is a 100-character limit preferred for documentation?

Good question.  Our style is generally one-sentence-per-line; that allows future diffs to be easier to read.  However, we do allow breaking a single sentence into multiple lines if necessary.  We don't have an exact number for line-length here, but I think 80 characters is too short.  Generally, my feeling is that if a particular sentence is too long to comfortably fit on one line, then usually the sentence is too long (there are exceptions to this, including when we are embedding URLs into the sentence, in which case the written sentence can get very long and we break it up).

I think the best thing to do is look at some of the other tutorials for reference.  While they aren't 100% consistent, that should give you a feeling about what we typically do.

abaeyens committed:
tutorial/integration_testing: apply suggestions from code review

Thanks for the many suggestions and feedback!
Applying the suggestions in multiple batches...

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Co-authored-by: Tomoya Fujita <Tomoya.Fujita@sony.com>
Signed-off-by: abaeyens <abaeyens@users.noreply.github.com>

abaeyens committed:
tutorial/integration_testing: apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: abaeyens <abaeyens@users.noreply.github.com>

abaeyens committed:
tutorial/integration_testing: apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Signed-off-by: abaeyens <abaeyens@users.noreply.github.com>

abaeyens committed:
tutorial/integration_testing: apply suggestions from code review

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Co-authored-by: Katherine Scott <katherineAScott@gmail.com>
Signed-off-by: abaeyens <abaeyens@users.noreply.github.com>

abaeyens committed:
tutorial/integration_testing: fix turtlesim capitalization

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
tutorial/integration_testing: remove "homework" section and trailing whitespace

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
tutorial/integration_testing: delete include launch file ref + split up sentence

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
tutorial/integration_testing: move test running and examination to existing tutorial

Suggestion, or prefer to put this somewhere else?

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
tutorial/integration_testing: fix whitespace

Follow https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
tutorial/integration_testing: link to launch_testing_ros package

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
tutorial/integration_testing: fix path

Should cover all packages, not just the 'app' package which may not even exist in the user's workspace.

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

clalancette commented on source/Tutorials/Intermediate/Testing/CLI.rst:
```diff
@@ -17,21 +17,58 @@ To compile and run the tests, simply run the `test <https://colcon.readthedocs.i
 :ref:`Sourcing the workspace <colcon-tutorial-source-the-environment>` before testing should not be necessary.
 ``colcon test`` makes sure that the tests run with the right environment, have access to their dependencies, etc.
 
+The above command suppresses the test output and exposes little about which tests succeed and which fail.
+Therefore while developing tests the ``--event-handlers`` option is useful to print all test output while the tests are running:
+
+.. code-block:: console
+
+  colcon test --event-handlers console_direct+
+
```
This particular change doesn't really have to do with this PR, so let's discuss this in a separate PR.

abaeyens replied:
Sure, addressed in https://github.com/ros2/ros2_documentation/pull/4881/commits/620dbab56d8c479b98b220f1bb5ba143e3aaf1f2.

clalancette commented on source/Tutorials/Intermediate/Testing/CLI.rst:
```diff
@@ -17,21 +17,58 @@ To compile and run the tests, simply run the `test <https://colcon.readthedocs.i
 
-To see the results, simply run the `test-result <https://colcon.readthedocs.io/en/released/reference/verb/test-result.html>`__ verb from ``colcon``.
+To see the results, simply run the `test-result <https://colcon.readthedocs.io/en/released/reference/verb/test-result.html>`__ verb from ``colcon``:
 
 .. code-block:: console
 
-  colcon test-result --all
+  $ colcon test-result --all
+  build/app/Testing/20241013-0810/Test.xml: 1 tests, 0 errors, 1 failure, 0 skipped
+  build/app/test_results/app/test_test_myapp.py.xunit.xml: 3 tests, 0 errors, 1 failure, 0 skipped
+
+  Summary: 4 tests, 0 errors, 2 failures, 0 skipped
+
+In the example above, the command lists two files:
+
+* a ctest-formatted XML file (a result of the ``CMakeLists.txt``)
+* an XUnit-formatted XML file, which is suitable for automatic report generation in automated testing in CI/CD pipelines.
```
Same with these changes.

clalancette commented on source/Tutorials/Intermediate/Testing/CLI.rst:
```diff
@@ -17,21 +17,58 @@ To compile and run the tests, simply run the `test <https://colcon.readthedocs.i
 
   colcon test-result --all --verbose
 
+A suitable non-ROS specific tool to visualize them all together is the
+`NodeJS package Xunit Viewer <https://github.com/lukejpreston/xunit-viewer>`_.
+It converts the XUnit files to HTML or straight into the terminal.
+For example, command and response (without highlighting):
+
+.. code-block:: console
+
+  $ xunit-viewer -r build -c
+    app.test_integration.launch_tests
+        ✗ test_publishes_pose time=0.52
+          - Traceback (most recent call last):
+    File "/home/user/ros_workspace/src/app/test/test_myapp.py", line 67, in test_publishes_pose
+        assert len(msgs_rx) > 100
+            ^^^^^^^^^^^^^^^^^^
+    AssertionError
+        ✓ test_exit_codes time=0.0
+        ✓ test_logs_spawning time=0.197
+
+  1 failure, 2 passed
+  Written to: /home/user/ros_workspace/index.html
+
```
And the same with these changes.

clalancette commented on source/Tutorials/Intermediate/Testing/Integration.rst:
```diff
@@ -0,0 +1,286 @@
+The package `launch_testing_ros <https://docs.ros.org/en/{DISTRO}/p/launch_testing_ros/index.html>`_ provides some convenience functions to achieve similar behavior,
+such as `WaitForTopics <https://docs.ros.org/en/{DISTRO}/p/launch_testing_ros/launch_testing_ros.wait_for_topics.html>`_.
+
+If you want to go further, you can implement a third test that publishes a twist message, asking the turtle to move, and subsequently checks that it moved by asserting that the pose message changed,
+effectively automating part of the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`.
```
```suggestion
If you want to go further, you can implement a third test that publishes a twist message, asking the turtle to move, and subsequently checks that it moved by asserting that the pose message changed.
This effectively automates part of the `Turtlesim introduction tutorial <../../Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim>`.
```

clalancette reviewed (changes requested):
A few more things to fix up.  This is getting closer to being merged.

abaeyens committed:
tutorial/integration_testing: Leave changes to CLI tutorial for other PR.

Signed-off-by: Arne Baeyens <mail@arnebaeyens.com>

abaeyens committed:
Update source/Tutorials/Intermediate/Testing/Integration.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>
Signed-off-by: abaeyens <abaeyens@users.noreply.github.com>

clalancette committed:
Minor formatting fixes.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette reviewed (approved):
This looks great, thank you for iterating.

abaeyens commented:
Thanks everyone for their efforts it getting this completed and merged, especially @clalancette! I'm glad I could contribute to the ROS documentation.
